Adding files for demographic analyses with selection (#30)
popsim-consortium/analysis2: n_t_with_s/README.md

# Readme for N(t) with selection example

This directory contains the code to run analysis of
demographic inference using multiple programs, with 
identical resulting data from any demographic model 
found in `stdpopsim` as input to each. 
This provides an example of how standardized 
population simulations may be used to compare methods.

The present analyses tests how demographic inference 
is influenced by selection.

The Snakemake workflow includes the necessary pipeline
for simulation, analyses, and plotting in an efficient manner.
Simply choose your parameters in a config file, 
and let snakemake handle the rest. 
(For large runs, the use of a cluster is highly encouraged)

> _NOTE_: This requires a bleeding edge install of `msprime` at the 
> moment. 

Reviewer: these lines can be cut
```suggestion
```

Author: done

## Workflow

The analysis includes three programs for predicting effective population 
size through time(`n_t`): 
[msmc](https://github.com/stschiff/msmc/issues/23),
[stairwayplot](https://sites.google.com/site/jpopgen/stairway-plot), and
[smc++](https://github.com/popgenmethods/smcpp).
There are four target rules that can be executed with the given parameters: 
`compound_msmc`,
`compound_smcpp`,
`compound_stairwayplot`, 
or you can run all three on the same simulated data with rule `all`.

To run an analysis, create a directory (wherever you want)
where all results, and intermediate
files will be stored. Next, create and place a file named `config.json` in it.
The json file must contain key : value combos described below. An example 
might look like this:

```json
{
    "seed" : 12345,
    "population_id" : 0,
    "num_samples_per_population" : [20, 0, 0],
    "num_sampled_genomes_msmc" : "2,8",
    "num_msmc_iterations" : 20,
    "replicates" : 10,
    "species" : "HomSap",
    "model" : "OutOfAfrica_3G09",
    "genetic_map" : "HapmapII_GRCh37",
    "chrm_list" : "chr22,chrX",

>     "mask_file" : "masks/HapmapII_GRCh37.mask.bed"

Reviewer: lets set the annotation and the DFE here in the config file. this will make the interface easier, e.g., 
```suggestion
    "mask_file" : "masks/HapmapII_GRCh37.mask.bed"
    "dfe_id": "Gamma_K17"
    "annotation_id": "ensembl_havana_104_exons"
```

Author: done

know about your cluster. We have provided an example of 
such a file in `cluster_talapas.json` that is for use with a
University of Oregon SLURM cluster. 

```json
{
    "__default__" :
    {
        "time" : "02:30:00",
        "n" : 1,
        "partition" : "kern,preempt",
        "mem" : "16G",
        "cores" : "4",
    },
    "run_msmc" :
    {
        "time" : "05:00:00",
        "mem" : "32G",
        "cores" : "4",
    }
    
}
```

At a minimum, you should
edit the names of the partition to match those on your own HPC.
The workflow can then be launched with the call

`snakemake -j 999 --config config="/projects/kernlab/jgallowa/homo_sapiens_Gutenkunst_0" --cluster-config cluster_talapas.json --cluster "sbatch -p {cluster.partition} -n {cluster.n} -t {cluster.time} --mem-per-cpu {cluster.mem} -c {cluster.cores}"`

(it may prove useful to simply put this command in a bash script)

and jobs will be automatically farmed out to the cluster

### Final output
The current final output is a plot comparing stairwayplot and smc++ estimates of $N(t)$, i.e., 
`homo_sapiens_Gutenkunst/all_estimated_Ne_{selection}.png` , 
with selection being 'none', 'all' or applied to the 'exons'.

> ## Parameter Description

Reviewer: might be a little more specific about where this params are? e.g., 
```suggestion
## Config file Parameter Description
```

Author: added change

run

`output_dir` : `<class 'str'>` This where all the intermediate files and results
will be stored. This includes all plots produced, as well as all input/output files
from the simulations and software runs organized into subdirectories by
replicate seed.

`species` : `<class 'module'>` from `stdpopsim` such as `stdpopsim.homo_sapiens`.
Chromosomes, demographics, and recombination maps should be derived from this.

`model` : `<class childclass Model>` 
This is the class that defines the demography associated with a species. All models
should inherit from `<class Model>`

`genetic_map` : `<class childclass GeneticMap>` This will define the genetic map
used for simulations.

`chrm_list` : `<class 'str'>` A string of the chromosome names you would like to simulate,
separated by commas. All chromosomes simulated will be fed
as a single input into each analysis by the inference programs, for each replicate.
Set to "all" to simulate all chromosomes for the genome.

`mask_file` : `<class 'str'>` A string indicating the path to a bed formatted file with
to be masked out of the analysis.

`dfe_id`: `<class 'str'>` A string indicating what Distribution of Fitness Effects to use.

> See http://sesame.uoregon.edu/~adkern/stdpopsim/doc/catalog.html to check all the possible DFEs

Reviewer: ```suggestion
See https://popsim-consortium.github.io/stdpopsim-docs/latest/catalog.html# to check all the possible DFEs
```

Author: added change

`selection`: `<class 'str'>` A string indicating where the DFE should be applied. So far three

> options are allowed: 'exons', where selection is imposed on exons, 'none', and 'all' –– in which 

Reviewer: so I don't think you need the `selection` parameter and the `annotation_id` in the config.yaml. Why not just allow for `annotation_id` to take the values `all` or `none` as well and get rid of this one?

Author: I will take out the annotation_id part but will leave the selection one. It makes it easier to run each pipeline of selection separately.

Reviewer: well I think we want the annotation_id though, right? like imagine we want to run this on different annotations

Author: I got it!

selection is insert in the entire contig under subject. 

`annotation_id`:  `<class 'str'>` A string indicating the annotation file to be used. For humans,

> the two annotations available are: ensembl_havana_104_exons and ensembl_havana_104_CDS.

Author: added change.
